src/haddock/gear/start_from_copy.py

```diff
@@ -0,0 +1,201 @@
+"""
+Copy and start run from copy gear.
+
+Contains the functionalities used in `haddock3-copy` CLI and in
+`--start-from-copy` flag for `haddock3` CLI.
+"""
+import shutil
+from pathlib import Path
+
+from haddock import log
+from haddock.core.defaults import MODULE_IO_FILE
+from haddock.gear.zerofill import zero_fill
+from haddock.libs.libontology import ModuleIO
+from haddock.libs.libworkflow import Workflow
+from haddock.modules import get_module_steps_folders
+
+
+START_FROM_COPY_DEFAULT = None
+
+
+class WorkflowManagerCopy:
+    """Read and execute workflows from copy."""
+
+    def __init__(self, workflow_params, start=0, **other_params):
+        self.start = start
+        self.recipe = Workflow(workflow_params, start=start, **other_params)
+
+    def run(self):
+        """High level workflow composer."""
+        for step in self.recipe.steps:
+            step.execute()
+
+
+def add_start_from_copy(parser):
+    """Add option to start-from-dir."""
+    parser.add_argument(
+        '--start-from-copy',
+        help=(
+            "Start a run from a run directory previously prepared with "
+            "the `haddock3-copy` CLI. Provide the run directory created "
+            "with `haddock3-copy` CLI."
+            ),
+        default=START_FROM_COPY_DEFAULT,
+        type=Path,
+        )
+
+
+def read_num_molecules_from_folder(folder):
+    """
+    Read the number of molecules from the first step folder.
+
+    1. Find the lower indexed step folder in `folder`.
+    2. Read the `io.json` file.
+    3. Count the number of items in the "output" key.
+    4. The above is the number of molecules.
+
+    Parameters
+    ----------
+    folder : Path or string
+        The run directory.
+
+    Returns
+    -------
+    int
+        The number of molecules found.
+    """
+    parent = Path(folder).resolve()
+    previous = get_module_steps_folders(folder)
+
+    io = ModuleIO()
+    previous_io = Path(parent, previous[0], MODULE_IO_FILE)
+    io.load(previous_io)
+
+    return len(io.output)
+
+
+def copy_renum_step_folders(indir, destdir, steps):
+    """
+    Copy step folders renumbering them sequentially in the run directory.
+
+    The content of the files is not modified.
+    See :py:func:`rename_step_contents`.
+
+    py:`gear.zerofill.zero_fill`: must be previously calibrated.
+
+    Example
+    -------
+    >>> steps = ['0_topoaa', '4_flexref']
+    >>> zero_fill.set_zerofill_number(len(steps))
+    >>> copy_renum_step_folders('run1', 'newrun', steps)
+
+    The initial structure::
+
+        run1/
+            1_rigidbody/
```

Comment:
Do I read it wrong if I think that `0_topoaa` is missing here in the initial structure?